kong/cmd/utils/prefix_handler.lua

```diff
@@ -378,6 +381,67 @@ local function write_process_secrets_file(path, data)
   return true
 end
 
+local function pre_create_lmdb(conf)
+  local LMDB_DIRECTORY = kong_constants.LMDB_DIRECTORY
+
+  local prefix = conf.prefix
+  local lmdb_path = conf.lmdb_environment_path or LMDB_DIRECTORY
+
+  local dir_name
+  if lmdb_path:sub(1, 1) == "/" then
+    dir_name = lmdb_path
+
+  else
+    dir_name = prefix .. "/" .. lmdb_path
+  end
+
+  if pl_path.isdir(dir_name) then
+    return true
+  end
+
+  local user, group = match(conf.nginx_user or "", "(%w+)%s*(%w*)")
+
+  if not user then
+    log.warn("nginx_user is not set in Kong config, "..
+             "skipping verification of LMDB file permissions")
+    return true
+  end
+
+  if not group or #group == 0 then
+    group = user
+  end
+
+  log.debug("LMDB directory '%s' does not exist, " ..
+            "pre-creating with the correct permissions",
+            dir_name)
+
+  local ok, err = pl_path.mkdir(dir_name)
+  if not ok then
+    return nil, "can not create directory for LMDB " .. dir_name ..
+                ", err: " .. err
+  end
+
+  local cmds = {
+    format("chown %s:%s %s && chmod 0700 %s",
+           user, group, dir_name, dir_name),
+
+    format("touch %s && chmod 0600 %s",
+           dir_name .. "/data.mdb", dir_name .. "/data.mdb"),
+
+    format("touch %s && chmod 0600 %s",
+           dir_name .. "/lock.mdb", dir_name .. "/lock.mdb"),
+  }
```

Comment:
what happens on systems that don't have these commands available?

Would this be better handled by [after-install.sh](https://github.com/Kong/kong-build-tools/blob/master/after-install.sh) ?

Reply:
I think for many cases after-install.sh _would_ be sufficient, but this is in $KONG_PREFIX, so its location can't really be 100% known at install time:

```shell
docker run \
  -e KONG_PREFIX=/foo \
  -v "$PWD:/foo" \
  my-kong-image
```

I agree on the principle though. We should only resort to shell/exec when absolutely necessary. Lua `io.*`, [LFS](https://lunarmodules.github.io/luafilesystem/manual.html), or FFI are all available to use.

The [file-log handler](https://github.com/Kong/kong/blob/8fda66e41db3399b345f2d211011d39ea4b09035/kong/plugins/file-log/handler.lua) has an example that can be followed.

For cases where we have to resort to shell, [resty.shell](https://github.com/openresty/lua-resty-shell) should be used because it eliminates the need for escaping/quoting arguments:

```lua
-- this is error-prone and unsafe
pl_utils.executeex(string.format("touch %s", "filename with spaces.txt"))
pl_utils.executeex(string.format("touch %s", "file.txt; rm -rf /"))

-- this is better
resty_shell.run({ "touch", "filename with spaces.txt" })
resty_shell.run({ "touch", "file.txt; rm -rf /" })
```

Reply:
Here I just follow the other test case's pattern, should we change them to `resty_shell`?

Reply:
`resty.shell` is probably fine

Reply:
does it make sense to use approaches like this (found in our private fork) to get rid of shell commands completely?

```lua
     # 2.7.1.2: kong/clustering/data_plane.lua#L191
      -- CONFIG_CACHE does not exist, pre create one with 0600 permission
      local fd = ffi.C.open(self.config_cache, bit.bor(system_constants.O_RDONLY(),
                                                       system_constants.O_CREAT()),
                                               bit.bor(system_constants.S_IRUSR(),
                                                       system_constants.S_IWUSR()))
      if fd == -1 then
        ngx_log(ngx_ERR, _log_prefix, "unable to pre-create cached config file: ",
                ffi.string(ffi.C.strerror(ffi.errno())))

      else
        ffi.C.close(fd)
      end
```